relayer/src/tx/connection.rs

```diff
@@ -1,72 +1,64 @@
-use crate::chain::{Chain, CosmosSDKChain};
-use crate::config::ChainConfig;
-use crate::error::{Error, Kind};
-use crate::keyring::store::{KeyEntry, KeyRingOperations};
+use prost_types::Any;
+use serde_json::Value;
+use std::str::FromStr;
+
 use bitcoin::hashes::hex::ToHex;
+
+use tendermint::account::Id as AccountId;
+use tendermint_rpc::Id;
+
 use ibc::ics03_connection::connection::Counterparty;
 use ibc::ics03_connection::msgs::conn_open_init::MsgConnectionOpenInit;
 use ibc::ics23_commitment::commitment::CommitmentPrefix;
 use ibc::ics24_host::identifier::{ClientId, ConnectionId};
 use ibc::tx_msg::Msg;
-use prost_types::Any;
-use serde_json::Value;
-use std::str::FromStr;
-use tendermint::account::Id as AccountId;
-use tendermint_rpc::Id;
+use ibc_proto::ibc::core::connection::v1::MsgConnectionOpenInit as RawMsgConnectionOpenInit;
+
+use crate::chain::{Chain, CosmosSDKChain};
+use crate::config::ChainConfig;
+use crate::error::{Error, Kind};
+use crate::keyring::store::{KeyEntry, KeyRingOperations};
 
 #[derive(Clone, Debug)]
 pub struct ConnectionOpenInitOptions {
-    pub src_client_id: ClientId,
-    pub dest_client_id: ClientId,
-    pub src_connection_id: ConnectionId,
-    pub dest_connection_id: Option<ConnectionId>,
-    pub src_chain_config: ChainConfig,
     pub dest_chain_config: ChainConfig,
-    pub signer_key: String,
+    pub src_chain_config: ChainConfig,
+    pub dest_client_id: ClientId,
+    pub src_client_id: ClientId,
+    pub dest_connection_id: ConnectionId,
+    pub src_connection_id: Option<ConnectionId>,
+    pub signer_seed: String,
     pub account_sequence: u64,
 }
 
-pub fn conn_init(opts: ConnectionOpenInitOptions) -> Result<Vec<u8>, Error> {
-    // Get the destination chain
-    let mut dest_chain = CosmosSDKChain::from_config(opts.clone().dest_chain_config)?;
+pub fn conn_init(opts: &ConnectionOpenInitOptions) -> Result<Vec<u8>, Error> {
+    // Get the source and destination chains
+    let src_chain = CosmosSDKChain::from_config(opts.src_chain_config.clone())?;
+    let mut dest_chain = CosmosSDKChain::from_config(opts.dest_chain_config.clone())?;
+
+    // Check that the destination chain will accept the message, i.e. it does not have the connection
+    dest_chain.check_connection_for_init(opts.dest_connection_id.clone())?;
```

Comment:
This method is quite fine-grained to keep in the chain trait. It would make sense to rely on a more generic method – simply `query_connection` – to implement this check, but let's consider this in a follow-up PR!

Reply:
Yes it is. I made this change after looking into the `OpenTry` message. Checking if the connection state is compatible with a certain handshake step is more complex for some messages, especially with flexible connection id. I think we will revise these methods once we have all connection tx-es implemented.

Reply:
Ok i removed this for now and also took out the message builder from chain.